server/scheduler.go

```diff
@@ -238,23 +238,36 @@ func (s *shuffleRegionScheduler) Schedule(cluster *clusterInfo) Operator {
 
 func newAddPeer(region *regionInfo, peer *metapb.Peer) Operator {
 	addPeer := newAddPeerOperator(region.GetId(), peer)
-	return newRegionOperator(region, addPeer)
+	return newRegionOperator(region, regionKind, addPeer)
 }
 
 func newRemovePeer(region *regionInfo, peer *metapb.Peer) Operator {
 	removePeer := newRemovePeerOperator(region.GetId(), peer)
-	return newRegionOperator(region, removePeer)
+	if region.Leader != nil && region.Leader.GetId() == peer.GetId() {
+		for _, newLeader := range region.Peers {
+			if newLeader.GetId() != region.Leader.GetId() {
+				transferLeader := newTransferLeaderOperator(region.GetId(), region.Leader, newLeader)
+				return newRegionOperator(region, regionKind, transferLeader, removePeer)
+			}
+		}
+		return nil
+	}
+	return newRegionOperator(region, regionKind, removePeer)
 }
 
 func newTransferPeer(region *regionInfo, oldPeer, newPeer *metapb.Peer) Operator {
 	addPeer := newAddPeerOperator(region.GetId(), newPeer)
 	removePeer := newRemovePeerOperator(region.GetId(), oldPeer)
```

Comment:
`newRemovePeer` returns a `regionOperator`, we will have to extract the `transferLeaderOperator` from it... I decided to do it straightforwardly by trying to find a follower. Please check it out.